src/chapters/aspdotnet-model-validation/studio.rst

```diff
@@ -0,0 +1,120 @@
+Studio: Spa User Validation
+===========================
+
+We’ll build on the :ref:`User Signup <user-signup-studio>` studio from last
+class, adding in model validation.
+
+Getting Started
+---------------
+
+Open up your ``SpaDay`` application and checkout the `user-signup-pt2 <https://github.com/LaunchCodeEducation/SpaDay/tree/user-signup-pt2>`__ branch. 
+
+Creating a New ViewModel
+------------------------
+
+To get started with adding validation to our application, we need to first make a ViewModel.
+After adding a folder for storing our ViewModels, create a new ViewModel for working with form submission called ``AddUserViewModel``.
+Add properties for the user's username, password, and email.
+Also, add a new property called ``VerifyPassword``.
+
+Add Validation Attributes
+-------------------------
+
+Navigate to the ``AddUserViewModel`` model class. Add `validation
+attributes <https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.dataannotations?view=netcore-3.1>`__
+to ensure these conditions are satisfied:
+
+#.  Username, password, and ``VerifyPassword`` are required.
+#.  Username is between 5 and 15 characters.
+#.  Email is optional.
+#.  If provided, the email has the format of a valid email address.
+#.  The password is between 6 and 20 characters.
+
+Refactoring ``UserController``
+------------------------------
+
+With our new ViewModel set up and ready, we need to refactor our action methods to make use of ``AddUserViewModel``.
+First, pass an instance of ``AddUserViewModel`` to ``User/Add.cshtml`` via our ``Add()`` action method.
+Next, to refactor the ``SubmitAddUserForm()`` action method, we need to do the following:
+
+#. Take in the instance of ``AddUserViewModel`` as a parameter.
+#. Use ``ModelState.IsValid`` to make sure that the conditions outlined using validation attributes have been met.
+#. If ``ModelState.IsValid`` is ``true``, we want to double check that the value of ``Password`` is equal to ``VerifyPassword``.
+   If it is, we can create an instance of ``User`` and pass it to the ``User/Index.cshtml`` view.
+   If the two passwords do not match, we need to reload the form with an error message displayed.
+
+Validating That Passwords Match
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Last studio, we added some validation checks to make sure the password fields match.
+Now we have two validation sections: one for the attribute-configured
+validation (which checks ``ModelState.IsValid``), and one that checks
+that the password fields match. Make sure they work in-sync with each
+other to properly return to the form if any of the validation conditions
+fail.
+
+.. admonition:: Tip
+
+   You can, in fact, validate that passwords match using attributes by
+   taking a slightly different approach than we’ve done here. We
+   outline how to do so in the Bonus Mission section.
+
+Refactoring the Views
```

Comment:
since both of these views depend on a model object, things get wonky if you hit those routes without the object getting passed in.
i think we can add a conditional to check for the presence on add and index

Reply:
I think we can add a conditional! I am going to note that as a possible improvement for later because the way I am seeing it done properly online seems to be beyond the scope of the class. I might have missed something though.